jaxlib/pytree.cc

```diff
@@ -0,0 +1,607 @@
+}
+
+template <typename H>
+H AbslHashValue(H h, const PyTreeDef& t) {
+  return H::combine_contiguous(std::move(h), t.traversal_.data(),
+                               t.traversal_.size());
+}
+
+bool PyTreeDef::operator==(const PyTreeDef& other) const {
+  if (traversal_.size() != other.traversal_.size()) {
+    return false;
+  }
+  for (size_t i = 0; i < traversal_.size(); ++i) {
+    const Node& a = traversal_[i];
+    const Node& b = other.traversal_[i];
+    if (a.kind != b.kind || a.arity != b.arity ||
+        (a.node_data.ptr() == nullptr) != (b.node_data.ptr() == nullptr) ||
+        a.custom != b.custom) {
+      return false;
+    }
+    if (a.node_data && a.node_data.not_equal(b.node_data)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+void PyTreeDef::FlattenHelper(py::handle handle, py::list* leaves,
+                              PyTreeDef* tree) {
+  Node node;
+  int start_num_nodes = tree->traversal_.size();
+  int start_num_leaves = leaves->size();
+  if (py::isinstance<py::none>(handle)) {
+    node.kind = Kind::kNone;
+  } else if (PyTuple_CheckExact(handle.ptr())) {
+    py::tuple tuple = py::reinterpret_borrow<py::tuple>(handle);
+    node.kind = Kind::kTuple;
+    node.arity = tuple.size();
+    for (py::handle entry : tuple) {
+      FlattenHelper(entry, leaves, tree);
+    }
+  } else if (PyList_CheckExact(handle.ptr())) {
+    py::list list = py::reinterpret_borrow<py::list>(handle);
+    node.kind = Kind::kList;
+    node.arity = list.size();
+    for (py::handle entry : list) {
+      FlattenHelper(entry, leaves, tree);
+    }
+  } else if (PyDict_CheckExact(handle.ptr())) {
+    py::dict dict = py::reinterpret_borrow<py::dict>(handle);
+    py::list keys = py::reinterpret_steal<py::list>(PyDict_Keys(dict.ptr()));
+    if (PyList_Sort(keys.ptr())) {
+      throw std::runtime_error("Dictionary key sort failed.");
+    }
+    for (py::handle key : keys) {
+      FlattenHelper(dict[key], leaves, tree);
+    }
+    node.kind = Kind::kDict;
+    node.arity = dict.size();
+    node.node_data = std::move(keys);
+  } else if ((node.custom = CustomNodeRegistry::Lookup(handle.get_type()))) {
+    node.kind = Kind::kCustom;
+    py::tuple out = node.custom->to_iterable(handle);
+    if (out.size() != 2) {
+      throw std::runtime_error(
+          "PyTree custom to_iterable function should return a pair");
+    }
+    node.node_data = out[1];
+    node.arity = 0;
+    for (py::handle entry : py::cast<py::iterable>(out[0])) {
+      ++node.arity;
+      FlattenHelper(entry, leaves, tree);
+    }
+  } else if (py::isinstance<py::tuple>(handle) &&
+             py::hasattr(handle, "_fields")) {
+    // We can only identify namedtuples heuristically, here by the presence of
+    // a _fields attribute.
+    py::tuple tuple = py::reinterpret_borrow<py::tuple>(handle);
+    node.kind = Kind::kNamedTuple;
+    node.arity = tuple.size();
+    node.node_data = py::reinterpret_borrow<py::object>(tuple.get_type());
```

Comment:
Do you need to get your own reference here?

Reply:
My understanding of `py::reinterpret_borrow` is that it acquires a reference. (As opposed to `py::reinterpret_steal`, which does not.)